### PR TITLE
Pv2 2394 providers requiring resubmission function

### PR DESCRIPTION
**Pull request by carlbm**

### CODE REVIEW

File: src/SFA.DAS.Payments.PeriodEnd.Application/Services/SubmissionJobsService.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+                                                            throw new ArgumentNullException(nameof(providersRequiringResubmissionRepository));
+        }
+
+        public async Task<SubmissionJobs> SuccessfulSubmissions()
+        {
+            var providers = await repository.GetLatestSuccessfulJobs();
+
+            var providersToExclude = await providersRequiringResubmissionRepository.GetAll();
+
+            providers.RemoveAll(x => providersToExclude.Contains(x.Ukprn));
```
Comment (tom-gough) [inline src/SFA.DAS.Payments.PeriodEnd.Application/Services/SubmissionJobsService.cs:32]: nit picking slightly again but would a .Where be more readable here?

providers.Where(x => providersToExclude.Any(y => y == x.Ukprn))

Just personal preference really but wondered what you think?

**Review by tom-gough — DISMISSED** [review]

File: src/SFA.DAS.Payments.PeriodEnd.Application/Repositories/PeriodEndRepository.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    public class PeriodEndRepository : IPeriodEndRepository
+    {
+        private readonly IPeriodEndDataContext dataContext;
+        private readonly IPaymentLogger logger;
+
+        public PeriodEndRepository(IPeriodEndDataContext dataContext, IPaymentLogger logger)
+        {
+            this.dataContext = dataContext ?? throw new ArgumentNullException(nameof(dataContext));
+            this.logger = logger ?? throw new ArgumentNullException(nameof(logger));
```
[the remark this answers is not in the document]
Comment (carlbm) [reply src/SFA.DAS.Payments.PeriodEnd.Application/Repositories/PeriodEndRepository.cs:25]: Added logging

File: src/SFA.DAS.Payments.PeriodEnd.Application/Services/SubmissionJobsService.cs (hunk 1 above)
Comment (carlbm) [reply src/SFA.DAS.Payments.PeriodEnd.Application/Services/SubmissionJobsService.cs:32]: I thought that as we're removing the items in the list it made more sense to use the `Remove` variant over the `Where`.

File: src/SFA.DAS.Payments.PeriodEnd.Function/Infrastructure/IoC/Modules/FunctionsModule.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+            builder.RegisterType<PeriodEndRepository>().As<IPeriodEndRepository>().InstancePerLifetimeScope();
+            builder.RegisterType<ProvidersRequiringReprocessingRepository>().As<IProvidersRequiringReprocessingRepository>().InstancePerLifetimeScope();
+
+            builder.Register((c, p) =>
+                {
+                    var config = c.Resolve<ISubmissionMetricsConfiguration>();
+                    var dbContextOptions = new DbContextOptionsBuilder<PeriodEndDataContext>()
+                        .UseSqlServer(config.PaymentsConnectionString, 
+                  optionsBuilder => optionsBuilder.CommandTimeout(270)).Options;
+                    return new PeriodEndDataContext(dbContextOptions);
```
Comment (hardik-desai-kcl) [inline src/SFA.DAS.Payments.PeriodEnd.Function/Infrastructure/IoC/Modules/FunctionsModule.cs:22]: in the other branch where we handle the apprenticeship updated events and submission successful events you have used the constructor with connection string and here you have used dbContextOptions constructor and reason why two different way of doing this?

File: src/SFA.DAS.Payments.PeriodEnd.Function/Infrastructure/Configuration/SubmissionMetricsConfiguration.cs [hunk 4]
```diff
@@ -0,0 +1,7 @@
+﻿namespace SFA.DAS.Payments.PeriodEnd.Function.Infrastructure.Configuration
+{
+    public class SubmissionMetricsConfiguration : ISubmissionMetricsConfiguration
```
Comment (hardik-desai-kcl) [inline src/SFA.DAS.Payments.PeriodEnd.Function/Infrastructure/Configuration/SubmissionMetricsConfiguration.cs:3]: ProvidersRequiringReprocessingConfiguration

File: src/SFA.DAS.Payments.PeriodEnd.Application/Services/SubmissionJobsService.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+using System.Threading.Tasks;
+
+namespace SFA.DAS.Payments.PeriodEnd.Application.Services
+{
+    public interface ISubmissionJobsService
+    {
+        Task<SubmissionJobs> SuccessfulSubmissions();
+    }
+
+    public class SubmissionJobsService : ISubmissionJobsService
```
Comment (hardik-desai-kcl) [inline src/SFA.DAS.Payments.PeriodEnd.Application/Services/SubmissionJobsService.cs:14]: ProvidersRequiringReprocessingService

File: src/SFA.DAS.Payments.PeriodEnd.AcceptanceTests/App.config [hunk 6]
```diff
@@ -29,6 +29,7 @@
     <add key="ApplicationInsightsInstrumentationKey" value="" />
     <add key="apiBaseUrl" value="" />
     <add key="ValidateDcAndDasServices" value="false" />
+    <add key="FunctionUri" value="" />
```
Comment (hardik-desai-kcl) [inline src/SFA.DAS.Payments.PeriodEnd.AcceptanceTests/App.config:32]: unless this is how we have done in the other projects could you please give this variable a good name?

**Review by hardik-desai-kcl — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.Payments.PeriodEnd.AcceptanceTests/App.config (hunk 6 above)
Comment (carlbm) [reply src/SFA.DAS.Payments.PeriodEnd.AcceptanceTests/App.config:32]: ✔️

File: src/SFA.DAS.Payments.PeriodEnd.Application/Services/SubmissionJobsService.cs (hunk 5 above)
Comment (carlbm) [reply src/SFA.DAS.Payments.PeriodEnd.Application/Services/SubmissionJobsService.cs:14]: ✔️

File: src/SFA.DAS.Payments.PeriodEnd.Function/Infrastructure/Configuration/SubmissionMetricsConfiguration.cs (hunk 4 above)
Comment (carlbm) [reply src/SFA.DAS.Payments.PeriodEnd.Function/Infrastructure/Configuration/SubmissionMetricsConfiguration.cs:3]: ✔️

File: src/SFA.DAS.Payments.PeriodEnd.Function/Infrastructure/IoC/Modules/FunctionsModule.cs (hunk 3 above)
Comment (carlbm) [reply src/SFA.DAS.Payments.PeriodEnd.Function/Infrastructure/IoC/Modules/FunctionsModule.cs:22]: We copied the existing pattern in both cases

Comment (carlbm) [reply src/SFA.DAS.Payments.PeriodEnd.Function/Infrastructure/IoC/Modules/FunctionsModule.cs]: Using the connection string constructor now

**Review by hardik-desai-kcl — APPROVED** [review]

**Review by hardik-desai-kcl — DISMISSED** [review]

**Review by hardik-desai-kcl — APPROVED** [review]

**Pull request merged**